src/Framework/PaymentGateways/Routes/GatewayRoute.php

```diff
@@ -2,77 +2,109 @@
 
 namespace Give\Framework\PaymentGateways\Routes;
 
+use Give\Framework\PaymentGateways\Actions\GenerateGatewayRouteUrl;
+use Give\Framework\PaymentGateways\Actions\ProcessOffsitePaymentRedirectOnGatewayRoute;
 use Give\Framework\PaymentGateways\Contracts\OffsiteGatewayInterface;
 use Give\Framework\PaymentGateways\DataTransferObjects\GatewayRouteData;
 use Give\Framework\PaymentGateways\Exceptions\PaymentGatewayException;
+use Give\Framework\PaymentGateways\Helpers\Gateway;
 use Give\Framework\PaymentGateways\PaymentGateway;
 use Give\Framework\PaymentGateways\PaymentGatewayRegister;
 
 /**
  * @since 2.18.0
+ *
+ * // Todo validate donation id before processing gateway method
+ * // Todo match gateway id with donation payment gateway id processing gateway method
  */
 class GatewayRoute
 {
+    /**
+     * @throws PaymentGatewayException
+     */
+    public function __invoke()
+    {
+        if ($this->isValidListener()) {
+            $data = GatewayRouteData::fromRequest($_GET);
+            $this->process($data);
+        }
+    }
+
     /**
      * @since 2.18.0
      *
      * @return void
      * @throws PaymentGatewayException
      */
-    public function __invoke()
+    public function process(GatewayRouteData $data)
     {
-        if ($this->isValidListener()) {
-            /** @var PaymentGatewayRegister $paymentGatewaysRegister */
-            $paymentGatewaysRegister = give(PaymentGatewayRegister::class);
-            $paymentGateways = $paymentGatewaysRegister->getPaymentGateways();
-            $gatewayIds = array_keys($paymentGateways);
+        /** @var PaymentGatewayRegister $paymentGatewaysRegister */
+        $paymentGatewaysRegister = give(PaymentGatewayRegister::class);
+        $paymentGateways = $paymentGatewaysRegister->getPaymentGateways();
+        $gatewayIds = array_keys($paymentGateways);
 
-            if (!$this->isValidRequest($gatewayIds)) {
-                throw new PaymentGatewayException('This route is not valid.');
-            }
+        if ( ! $this->isValidRequest($gatewayIds, $data->gatewayId)) {
+            throw new PaymentGatewayException('This route is not valid.');
+        }
 
-            $data = GatewayRouteData::fromRequest($_GET);
+        if ( ! $this->hasValidNonce($data)) {
+            throw new PaymentGatewayException('This route does not have valid nonce.');
+        }
 
-            /** @var PaymentGateway $gateway */
-            $gateway = give($paymentGateways[$data->gatewayId]);
+        /** @var PaymentGateway $gateway */
+        $gateway = give($paymentGateways[$data->gatewayId]);
 
-            $allowedGatewayMethods = $gateway->routeMethods;
+        $allowedGatewayMethods = Gateway::isOffsitePaymentGateway($gateway) ? $gateway->routeMethods : [];
```

Comment:
@ravinderk gateways do not need to implement the offsite interface to have routeMethods.  please revert this change.